core/src/main/java/hudson/model/DescriptorVisibilityFilter.java

```diff
@@ -57,15 +77,97 @@
+                        }
+                        continue OUTER; // veto-ed. not shown
+                    }
+                    if (!f.filter(context, d)) {
+                        if (LOGGER.isLoggable(Level.CONFIG)) {
+                            LOGGER.config("Filter " + f + " hides " + d + " in context " + context);
+                        }
+                        continue OUTER; // veto-ed. not shown
+                    }
+                } catch (Error e) {
+                    LOGGER.log(Level.WARNING, "Encountered error while processing filter " + f + " for context " + context, e);
+                    throw e;
+                } catch (Throwable e) {
+                    LOGGER.log(logLevelFor(f), "Uncaught exception from filter " + f + " for context " + context, e);
+                    continue OUTER; // veto-ed. not shown
+                }
+            }
+            r.add(d);
+        }
+        return r;
+    }
+
+    public static <T extends Descriptor> List<T> applyType(Class<?> contextClass, Iterable<T> source) {
+        ExtensionList<DescriptorVisibilityFilter> filters = all();
+        List<T> r = new ArrayList<T>();
+
+        OUTER:
+        for (T d : source) {
+            if (LOGGER.isLoggable(Level.FINE)) {
+                LOGGER.fine("Determining visibility of " + d + " in contexts of type " + contextClass);
+            }
+            for (DescriptorVisibilityFilter f : filters) {
+                if (LOGGER.isLoggable(Level.FINER)) {
+                    LOGGER.finer("Querying " + f + " for visibility of " + d + " in type " + contextClass);
+                }
+                try {
+                    if (contextClass != null && !f.filterType(contextClass, d)) {
+                        if (LOGGER.isLoggable(Level.CONFIG)) {
+                            LOGGER.config("Filter " + f + " hides " + d + " in contexts of type " + contextClass);
+                        }
+                        continue OUTER; // veto-ed. not shown
                     }
+                } catch (Error e) {
+                    LOGGER.log(Level.WARNING,
+                            "Encountered error while processing filter " + f + " for contexts of type " + contextClass, e);
+                    throw e;
+                } catch (Throwable e) {
+                    LOGGER.log(logLevelFor(f), "Uncaught exception from filter " + f + " for context of type " + contextClass, e);
                     continue OUTER; // veto-ed. not shown
                 }
             }
             r.add(d);
         }
         return r;
     }
+
+    /**
+     * Returns the {@link Level} to log an uncaught exception from a {@link DescriptorVisibilityFilter}. We
+     * need to suppress repeated exceptions as there can be many invocations of the {@link DescriptorVisibilityFilter}
+     * triggered by the UI and spamming the logs would be bad.
```

Comment:
Seems like overkill to me. It should be the responsibility of the logging system to avoid printing a massive number of messages from one component, especially if they are similar or identical. KISS in this code.